influxdb/2.0/entrypoint.sh

```diff
@@ -0,0 +1,267 @@
+        --org ${INFLUXDB_INIT_ORG}
+        --bucket ${INFLUXDB_INIT_BUCKET}
+    )
+    if [ -n "$INFLUXDB_INIT_RETENTION" ]; then
+        setup_args=(${setup_args[@]} --retention ${INFLUXDB_INIT_RETENTION})
+    fi
+
+    influx setup ${setup_args[@]}
+}
+
+# Get the IDs of the initial user/org/bucket created during setup, and export them into the env.
+# We do this to help with arbitrary user scripts, since many influx CLI commands only take IDs.
+function set_init_resource_ids () {
+    export INFLUXDB_INIT_USER_ID=$(influx user list -n ${INFLUXDB_INIT_USER} --hide-headers | cut -f 1)
+    export INFLUXDB_INIT_ORG_ID=$(influx org list -n ${INFLUXDB_INIT_ORG} --hide-headers | cut -f 1)
+    export INFLUXDB_INIT_BUCKET_ID=$(influx bucket list -n ${INFLUXDB_INIT_BUCKET} --hide-headers | cut -f 1)
+}
+
+# Execute all shell files mounted into the expected path for user-defined startup scripts.
+function run_user_scripts () {
+    if [ -d ${USER_SCRIPT_DIR} ]; then
+        log info "Executing user-provided scripts" script_dir ${USER_SCRIPT_DIR}
+        run-parts --regex ".*sh$" --report --exit-on-error ${USER_SCRIPT_DIR}
+    fi
+}
+
+# Perform initial setup on the InfluxDB instance, either by setting up fresh metadata
+# or by upgrading existing V1 data.
+function init_influxd () {
+    if [[ ${INFLUXDB_INIT_MODE} != setup && ${INFLUXDB_INIT_MODE} != upgrade ]]; then
+        log error "found invalid INFLUXDB_INIT_MODE, valid values are 'setup' and 'upgrade'" INFLUXDB_INIT_MODE ${INFLUXDB_INIT_MODE}
+        exit 1
+    fi
+    ensure_init_vars_set
+    trap "cleanup_influxd" EXIT
+
+    # The upgrade process needs to run before we boot the server, otherwise the
+    # boltdb file will be generated and cause conflicts.
+    if [ ${INFLUXDB_INIT_MODE} = upgrade ]; then
+        upgrade_influxd ${@}
+    fi
+
+    # Start influxd in the background.
+    log info "booting influxd server in the background"
+    influxd ${@} &
+    local -r influxd_pid="$!"
+    wait_for_influxd
+
+    # Use the influx CLI to create an initial user/org/bucket.
+    if [ ${INFLUXDB_INIT_MODE} = setup ]; then
+        setup_influxd
+    fi
+
+    set_init_resource_ids
+    run_user_scripts
+
+    log info "initialization complete, bringing influxd to foreground"
+    wait ${influxd_pid}
+}
+
+# Run influxd, with optional setup logic.
+function influxd_main () {
+    local -r bolt_path=$(influxd print-config --key-name bolt-path ${@})
+    local -r engine_path=$(influxd print-config --key-name engine-path ${@})
+
+    if test -f ${bolt_path}; then
+        log info "found existing boltdb file, skipping setup wrapper" bolt_path ${bolt_path}
+        influxd ${@}
+    elif [ -z "$INFLUXDB_INIT_MODE" ]; then
+        log warn "boltdb not found at configured path, but INFLUXDB_INIT_MODE not specified, skipping setup wrapper" bolt_path ${bolt_path}
+        influxd ${@}
+    else
+        init_influxd ${@}
+    fi
+}
+
+function main () {
```

Comment:
I don't understand this script very well, but the only recommendation is I don't see `exec` being used in some places. In general, the entrypoint.sh script should always end with calling `exec` to whatever the final command is. This allows `influxd` to be invoked with PID 1 and be the primary receiver of any signals. The shell can interfere with signal propagation.

https://stackoverflow.com/a/48096779/326250

Reply:
Hmmm in the setup/upgrade case, does that mean I'd need to spin up the server once for init logic, shut it down, then start it a 2nd time using `exec`? It's what the 1.x entrypoints do, but I was hoping to avoid it... I guess the alternative would be a bunch of ugly `trap` calls to forward signals we care about onto the `influxd` process